File: algorithmic_toolbox/week_2/01_introduction_starter_files/fibonacci_sum_last_digit/fibonacci_sum_last_digit.py
# Uses python3
import sys
import unittest


def fibonacci_sum_naive(n):
    if n <= 1:
        return n

    previous = 0
    current = 1
    sum = 1

    for _ in range(n - 1):
        previous, current = current, previous + current
        sum += current

    return sum % 10


def get_fibonacci_last_digit_fast(n):
    if n <= 1:
        return n

    previous = 0
    current = 1

    for _ in range(n - 1):
        previous, current = current % 10, (previous + current) % 10

    return current


def fibonacci_sum_fast(n):
    new_n = (n + 2) % 60
    new_last = get_fibonacci_last_digit_fast(new_n)
    if new_last == 0:
        return 9
    else:
        return new_last - 1


class MyTest(unittest.TestCase):
    def test_naive(self):
        self.assertEqual(fibonacci_sum_naive(3), 4)
        self.assertEqual(fibonacci_sum_naive(100), 5)

    def test_fast(self):
        self.assertEqual(fibonacci_sum_fast(3), 4)
        self.assertEqual(fibonacci_sum_fast(100), 5)


if __name__ == '__main__':
    # unittest.main()

    # while(True):
    #   a = random.randint(1, 100)

    #   res1 = fibonacci_sum_naive(a)
    #   res2 = fibonacci_sum_fast(a)
    #   if (res1 != res2):
    #     print('Generated: {}'.format(a))
    #     print('Wrond answer: {} {}'.format(res1, res2))
    #     break
    #   else:
    #     print('OK: {} {}'.format(res1, res2))

    input = sys.stdin.read()
    n = int(input)
    print(fibonacci_sum_fast(n))

File: algorithmic_toolbox/week_2/01_introduction_starter_files/fibonacci_sum_last_digit/fibonacci_sum_last_digit.cpp
#include <iostream>

int fibonacci_sum_naive(long long n) {
    if (n <= 1)
        return n;

    long long previous = 0;
    long long current = 1;
    long long sum = 1;

    for (long long i = 0; i < n - 1; ++i) {
        long long tmp_previous = previous;
        previous = current;
        current = tmp_previous + current;
        sum += current;
    }

    return sum % 10;
}

int get_fibonacci_last_digit_fast(long long n) {
    if (n <= 1)
        return n;

    int previous = 0;
    int current = 1;

    for (int i = 0; i < n - 1; ++i) {
        int tmp_previous = previous % 10;
        previous = current % 10;
        current = tmp_previous + current % 10;
    }
    return current % 10;
}

int fibonacci_sum_fast(long long n) {
    int new_n = (n + 2) % 60;
    int new_last = get_fibonacci_last_digit_fast(new_n);
    if (new_last == 0) {
        return 9;
    } else {
        return new_last - 1;
    }
}

int main() {
    long long n = 0;
    std::cin >> n;
    std::cout << fibonacci_sum_fast(n);
}
